File: Lab 13/13.8(4005).py
class Node:
    def __init__(self, item):
        self.mItem = item
        self.mNext = None


class Queue:
    def __init__(self):
        self.mFront = None
        self.mBack = None
        self._size  = 0

    def empty(self):
        return self.mFront is None

    def push(self, item):
        new_node = Node(item)
        if self.empty():
            self.mFront = new_node
        else:
            self.mBack.mNext = new_node
        self.mBack = new_node
        self._size += 1

    def pop(self):
        if self.empty():
            return 'error'
        current_front = self.mFront
        item = current_front.mItem
        self.mFront = self.mFront.mNext
        del current_front
        if self.mFront is None:
            self.mBack = None
        self._size -= 1
        return item

    def front(self):
        if self.empty():
            return 'error'
        return self.mFront.mItem

    def size(self):
        return self._size

first = Queue()
second = Queue()
i = 0

n = int(input())
for el in map(int, input().split()):
    first.push(el)
for el in map(int, input().split()):
    second.push(el)

while i < 2*10**5:
    i += 1
    first_c = first.pop()
    second_c = second.pop()
    if first_c == 0 and second_c == n-1:
        first.push(first_c)
        first.push(second_c)
    elif first_c == n-1 and second_c == 0:
        second.push(first_c)
        second.push(second_c)
    elif first_c > second_c:
        first.push(first_c)
        first.push(second_c)
    else:
        second.push(first_c)
        second.push(second_c)
    if first.empty():
        print(f'second {i}')
        break
    elif second.empty():
        print(f'first {i}')
        break

File: Lab 13/13/13/13.8(4005).cpp
#include <iostream> 
#include <string>
#include <queue>

using namespace std;

int main1(void) {
	int n, i, el, f_c, s_c;
	queue<int> first, second;
	cin >> n;
	for (i = 0;i < n / 2;i++) {
		cin >> el;
		first.push(el);
	}
	for (i = 0;i < n / 2;i++) {
		cin >> el;
		second.push(el);
	}
	i = 0;
	while (i<200000)
	{
		i++;
		f_c = first.front();
		first.pop();
		s_c = second.front();
		second.pop();
		if (f_c == 0 && s_c == n - 1) {
			first.push(f_c);
			first.push(s_c);
		}
		else if (f_c == n - 1 && s_c == 0) {
			second.push(f_c);
			second.push(s_c);
		}
		else if (f_c > s_c) {
			first.push(f_c);
			first.push(s_c);
		}
		else {
			second.push(f_c);
			second.push(s_c);
		}
		if (first.empty()) {
			cout << "second " << i;
			return 0;
		}
		else if (second.empty()) {
			cout << "first " << i;
			return 0;
		}
	}
	cout << "draw";
	return 0;
}